core/arch/arm/kernel/thread.c

```diff
@@ -1186,26 +1181,39 @@ void thread_rpc_free_arg(uint64_t cookie)
 	}
 }
 
-void thread_rpc_alloc_arg(size_t size, paddr_t *arg, uint64_t *cookie)
+struct mobj *thread_rpc_alloc_arg(size_t size, uint64_t *cookie)
 {
-	paddr_t pa;
+	paddr_t pa, page;
 	uint64_t co;
 	uint32_t rpc_args[THREAD_RPC_NUM_ARGS] = {
 		OPTEE_SMC_RETURN_RPC_ALLOC, size
 	};
+	struct mobj *mobj = NULL;
 
 	thread_rpc(rpc_args);
 
 	pa = reg_pair_to_64(rpc_args[1], rpc_args[2]);
 	co = reg_pair_to_64(rpc_args[4], rpc_args[5]);
-	if (!check_alloced_shm(pa, size, sizeof(uint64_t))) {
-		thread_rpc_free_arg(co);
-		pa = 0;
-		co = 0;
-	}
 
-	*arg = pa;
+	if (!ALIGNMENT_IS_OK(pa, struct optee_msg_arg))
+		goto err;
+	/* Check if this region is in static shared space */
```

Comment:
Thanks. Good idea. This fixed regression caused by v2 patch. @jenswi-linaro stepped right into it.